test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts

```diff
@@ -0,0 +1,257 @@
+      ERC721_WARPER_PRESET_IDS.ERC721_CONFIGURABLE_PRESET,
+      {
+        warperPresetFactory,
+      },
+    );
+    await Erc721ConfigurablePresetValidator.validateWarperState(
+      deployedWarperAddress,
+      originalCollection,
+      {
+        availabilityPeriodStart: 0,
+        availabilityPeriodEnd: MAX_TIME_PROTOCOL_UINT32,
+      },
+      { minRentalPeriod: 0, maxRentalPeriod: MAX_TIME_PROTOCOL_UINT32 },
+      { metahub, reader: stranger },
+    );
+    await WarperManagerValidator.validateWarperRegistration(
+      setupUniverseSuccessTx,
+      universeId,
+      deployedWarperAddress,
+      originalCollection.address,
+      ASSET_CLASS_IDS.ERC721,
+      { warperManager },
+    );
+    await TaxTermsRegistryValidator.validateUniverseWarperTaxTermsRegistration(
+      setupUniverseSuccessTx,
+      universeId,
+      deployedWarperAddress,
+      taxTerms,
+      { taxTermsRegistry },
+    );
+  });
+
+  it(`${TEST_CASE(TEST_OK)} Setup Universe with a pre-deployed Warper`, async () => {
+    const { deployedWarperAddress } = await new WarperDeployFromPresetBuilder(warperPresetFactory, universeOwner)
+      .withWarperConstructor(warperConstructor)
+      .deployFromPresetWithSuccess();
+    const universeBuilder = new UniverseWithErc721WarperBuilderV1(universeWizardV1, universeOwner)
+      .withUniverseConstructor(universeConstructor)
+      .withWarperConstructor(warperConstructor)
+      .withWarperAddress(deployedWarperAddress)
+      .withTaxTerms(taxTerms);
+    const { setupUniverseSuccessTx, universeId } = await universeBuilder.setupUniverseWithSuccessAndTx();
+
+    await UniverseValidatorV1.validateUniverseSetupProperly(
+      setupUniverseSuccessTx,
+      universeId,
+      universeOwner,
+      universeConstructor.getUniverseParamsSafely(),
+      {
+        universeRegistry,
+        reader: stranger,
+      },
+    );
+    await WarperManagerValidator.validateWarperRegistration(
+      setupUniverseSuccessTx,
+      universeId,
+      deployedWarperAddress,
+      originalCollection.address,
+      ASSET_CLASS_IDS.ERC721,
+      { warperManager },
+    );
+    await TaxTermsRegistryValidator.validateUniverseWarperTaxTermsRegistration(
+      setupUniverseSuccessTx,
+      universeId,
+      deployedWarperAddress,
+      taxTerms,
+      { taxTermsRegistry },
+    );
+  });
+
+  it(`${TEST_CASE(TEST_NOT_OK)} Setup Universe with an empty name`, async () => {
+    universeConstructor = universeConstructor.withUniverseName('');
+    const setupUniverseTx = new UniverseWithErc721WarperBuilderV1(universeWizardV1, universeOwner)
+      .withUniverseConstructor(universeConstructor)
+      .withWarperConstructor(warperConstructor)
+      .withTaxTerms(taxTerms)
+      .setupUniverseWithTxPromiseOnly();
+
+    await expect(setupUniverseTx).to.be.revertedWithCustomError(
+      { interface: IUniverseRegistry__factory.createInterface() },
+      UNIVERSE_REGISTRY_CUSTOM_ERRORS.EMPTY_UNIVERSE_NAME,
```

Comment:
At the moment when create this UNIVERSE_REGISTRY_CUSTOM_ERRORS, did not think about separating ,but would be great if you could level UNIVERSE_REGISTRY_CUSTOM_ERRORS to wht related to UniverseRegistry and move what related to IUniverseRegistry to IUNIVERSE_REGISTRY_CUSTOM_ERRORS 👍

Reply:
Umm, what is the difference? I mean, between UNIVERSE_REGISTRY and IUNIVERSE_REGISTRY? (solidity ultra beginner here).

Reply:
One is the actual contract (it is getting physically deployed on Blockchain) and another one is interface of this contract. This is just our convention that UniverseRegistry (and similar) have IUniverseRegistry (and similar) named interface. 

Since for example with custom errors - you must specifically provide typechain interface of contract/interface/library where the custom error is declared. And it just makes sense to separate errors and events by the place of their declaration 👍

Reply:
> would be great if you could level UNIVERSE_REGISTRY_CUSTOM_ERRORS to wht related to UniverseRegistry and move what related to IUniverseRegistry to IUNIVERSE_REGISTRY_CUSTOM_ERRORS

would this mean that we could have both UNIVERSE and IUNIVERSE errors? Not sure if I understand this. Seems to me that UNIVERSE.interface should have everything that is in IUNIVERSE? Or do you mean rename UNIVERSE errors to IUNIVERSE errors?

Reply:
When writing contracts, you can declare the errors either in contracts/abstract contracts/interfaces. Unfortunately, I think this is specifics of hardhat matchers that you must provide specifically the place where it is declared.

So if contract implements interface, here it does not mean that declaration is also inside the contract.

So that is why this difference is crucial.


Most of times we will have I<CONTRACT_NAME>_CUSTOM_ERRORS, since I think mostly we were trying to declare them in interfaces. As well as in libraries - which are separate <LIB_BAME>_CUSTOM_ERRORS. Sometimes inside the contract itself, making <CONTRACT_NAME>_CUSTOM_ERRORS

Reply:
Thanks for explanation, now it makes sense :) Will make an update!